src/plugins/api/index_pattern_endpoints.js

```diff
@@ -0,0 +1,221 @@
+        getMappings(pattern, client),
+        stitchPatternAndMappings
+      ).then(removeDeprecatedFieldProps)
+        .then(function (pattern) {
+          reply(pattern);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns',
+    method: 'POST',
+    handler: function (req, reply) {
+      const client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
+      const mappings = _.omit(_.mapValues(_.indexBy(req.payload.fields, 'name'), (value) => {
+        return value.mapping;
+      }), _.isUndefined);
+      indexPattern.fields = JSON.stringify(_.map(indexPattern.fields, (field) => {
+        return _.omit(field, 'mapping');
+      }));
+
+
+      client.create({
+        index: '.kibana',
+        type: 'index-pattern',
+        id: indexPattern.title,
+        body: indexPattern
+      }).then((patternResponse) => {
+        if (!isWildcard || _.isEmpty(mappings)) {
+          return patternResponse;
+        }
+        else {
+          return client.indices.exists({
+            index: indexPattern.title
+          }).then((matchingIndices) => {
+            if (matchingIndices) {
+              throw Boom.conflict('Cannot create an index template if existing indices already match index pattern');
+            }
+            else {
+              return client.indices.putTemplate({
+                order: 0,
+                create: true,
+                name: patternToTemplate(indexPattern.title),
+                body: {
+                  template: indexPattern.title,
+                  mappings: {
+                    _default_: {
+                      properties: mappings
+                    }
+                  }
+                }
+              });
+            }
+          }).catch((templateError) => {
+            return client.delete({
+              index: '.kibana',
+              type: 'index-pattern',
+              id: indexPattern.title
+            }).then(() => {
+              throw templateError;
+            }, () => {
+              throw new Error(`index-pattern ${indexPattern.title} created successfully but index template creation
+                failed. Failed to rollback index-pattern creation, must delete manually.`);
+            });
+          });
+        }
+      }).then(() => {
+        reply('success').statusCode = 201;
+      }).catch(function (error) {
+        reply(handleESError(error));
+      });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns/{id}',
+    method: 'PUT',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
```

Comment:
I see what you're saying now, the index template won't match anything if [] is used because ES doesn't understand it. I'm hesitant to give the user the ability to set the template pattern themselves because it would allow the template and index pattern to get out of sync. As I understand it the pattern and template should match all the same indices. Given that the `[]` syntax is deprecated anyways I'd lean towards only creating a template when `*` is used. @rashidkpc any thoughts on this?